### PR TITLE
feat: make voting easier with delegate modal

### PR DESCRIPTION
**Pull request by dated**
## Proposed changes

#803 

vote

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING](https://docs.ark.io/guidebook/contribution-guidelines/contributing.html) documentation
- [x] Lint and unit tests pass locally with my changes

### CODE REVIEW

**Comment by dated** [comment]
This PR depends on https://github.com/ArkEcosystem/desktop-wallet/pull/1023 and won't work as intended without those changes.

**Comment by dated** [comment]
There is one edge case which currently isn't handled correctly, i.e. when the the vote/unvote transaction gets broadcast and confirmed before the broadcast is finished. The vote details won't be loaded correctly in that case. If you want to try it out in the meantime @alexbarnsley 😉

**Comment by alexbarnsley** [comment]
🤔 	yeh will have a test either today or tomorrow 👌

**Comment by dated** [comment]
the above edge case is handled correctly now.

**Comment by j-a-m-l** [comment]
@dated is this already finished?

**Comment by dated** [comment]
If you can give it a try that'd be great. I'll have another look at it to see if somethings missing.

File: src/renderer/components/Input/InputDelegate.vue [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+        delegate.name = `${truncate(object.username, 25)} (${this.wallet_truncate(object.address)})`
+
+        return delegate
+      })
+
+      const results = _.orderBy(delegates, (object) => {
+        return object.name || object.address.toLowerCase()
+      })
+
+      return results.reduce((map, delegate, index) => {
```
Comment (j-a-m-l) [inline src/renderer/components/Input/InputDelegate.vue]: I'd use a different name for `map`; something like `acc`, `all`, `final`, `delegates`, etc.

**Comment by dated** [comment]
I'm unable to reproduce what you are saying unfortunately, the wallet shows as processing after the vote has been sent successfully until the actually confirmed vote transaction has been received.

**Comment by dated** [comment]
Thanks for that, I'll fix that ASAP.

**Comment by dated** [comment]
I'll mark this as WIP again, I want to properly refactor the vote handling as currently there is no persistent state of the yet to be processed votes.

**Comment by dated** [comment]
Can you have another look @j-a-m-l?

**Comment by dated** [comment]
No, it's an error caused by my changes during the loading of a profile into the session. I'll add proper storage of the unconfirmed vote(s) to the profile to fix this and to handle the case where the user would vote for a wallet, wait for the transaction to be sent and close the wallet before the actual confirmation has been received.

**Comment by j-a-m-l** [comment]
@dated there is a `TypeError`:

```
TypeError: Cannot read property 'filter' of undefined
    at Action._callee5$ (webpack-internal:///./src/renderer/services/synchronizer/wallets.js:439:77)
    at tryCatch (webpack-internal:///./node_modules/regenerator-runtime/runtime.js:62:40)
...
```

**Comment by dated** [comment]
For backwards compatibility, when loading a profile, the `unconfirmedVotes` are now assigned if not present.

**Comment by j-a-m-l** [comment]
@dated are you going to include here #1097?

**Comment by dated** [comment]
If this PR is fine I'd fix that in a separate PR @j-a-m-l.

File: src/renderer/store/modules/session.js [hunk 2]
```diff
@@ -186,6 +194,11 @@ export default {
       const profile = rootGetters['profile/byId'](profileId)
       if (!profile) return
 
+      if (!profile.unconfirmedVotes) {
+        profile.unconfirmedVotes = []
+        dispatch('profile/update', profile)
```
Comment (j-a-m-l) [inline src/renderer/store/modules/session.js]: `[vuex] unknown local action type: profile/update, global type: session/profile/update` on the dashboard

File: src/renderer/i18n/locales/en-US.js [hunk 3]
```diff
@@ -883,7 +894,11 @@ export default {
     VOTE_DELEGATE: 'Vote Delegate {delegate}',
     UNVOTE_DELEGATE: 'Unvote Delegate {delegate}',
     VOTED_FOR: 'You voted for delegate {delegate}',
-    WALLET_VOTED_FOR: 'This wallet voted for delegate {delegate}'
+    WALLET_VOTED_FOR: 'This wallet voted for delegate {delegate}',
+    NO_VOTE: 'You haven\'t voted yet, click on the button to the right to select a delegate',
```
Comment (j-a-m-l) [inline src/renderer/i18n/locales/en-US.js]: Maybe "This wallet is not voting. Click on any delegate or search it to vote"?

File: src/renderer/i18n/locales/en-US.js [hunk 4]
```diff
@@ -883,7 +894,11 @@ export default {
...
+    AWAITING_VOTE_CONFIRMATION: 'Your {type} is awaiting confirmation',
+    LOADING_VOTE: 'Loading vote details...',
+    SELECT_DELEGATE: 'Select delegate'
```
Comment (j-a-m-l) [inline src/renderer/i18n/locales/en-US.js]: "Search delegate"

**Comment by dated** [comment]
> I'd show the select/search button always because it could be used to find delegates

@j-a-m-l that makes sense. I'll work on it later.

File: src/renderer/i18n/locales/en-US.js (hunk 3 above)
Comment (dated) [reply src/renderer/i18n/locales/en-US.js]: Changed to `This wallet is not voting - use the search or click on any delegate to cast your vote`, does that sound ok?

Comment (j-a-m-l) [reply src/renderer/i18n/locales/en-US.js]: Uhm, I'm not completely sure; users could think that clicking on a delegate would cast their vote directly.

Something like this "This wallet is not voting. To view more information and vote, use the search or click on any delegate"?

**Comment by j-a-m-l** [comment]
Thanks, @dated

File: src/renderer/i18n/locales/en-US.js (hunk 3 above)
Comment (dated) [reply src/renderer/i18n/locales/en-US.js]:
```suggestion
    NO_VOTE: 'This wallet is not voting. To view more information and vote, use the search or click on any delegate',
```

File: src/renderer/components/Input/InputDelegate.vue [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+        return this.dropdownValue || this.inputValue
+      },
+      set (value) {
+        this.updateInputValue(value)
+        this.$emit('input', value)
+      }
+    },
+
+    suggestions () {
+      if (!this.currentProfile) return []
```
Comment (alexbarnsley) [inline src/renderer/components/Input/InputDelegate.vue]: Try to keep if statements consistent with curly brackets as it helps with readability

File: src/renderer/components/Input/InputDelegate.vue [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+      this.$nextTick(() => this.closeDropdown())
+    },
+
+    onFocus () {
+      this.isFocused = true
+      this.$emit('focus')
+    },
+
+    onEnter () {
+      if (!this.dropdownValue) return
```
Comment (alexbarnsley) [inline src/renderer/components/Input/InputDelegate.vue]: Try to keep if statements consistent with curly brackets as it helps with readability

File: src/renderer/components/Input/InputDelegate.vue [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+    },
+
+    async updateInputValue (value) {
+      this.inputValue = value
+      // Inform Vuelidate that the value changed
+      this.$v.model.$touch()
+    },
+
+    __setSuggestion (value) {
+      if (!this.hasSuggestions) return
```
Comment (alexbarnsley) [inline src/renderer/components/Input/InputDelegate.vue]: Try to keep if statements consistent with curly brackets as it helps with readability

File: src/renderer/components/Input/InputDelegate.vue [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+        this.$refs.input.setSelectionRange(this.inputValue.length, this.dropdownValue.length)
+      })
+    }
+  },
+
+  validations: {
+    model: {
+      required,
+      isValid (value) {
+        return !!this.$store.getters['delegate/byUsername'](value) ||
```
Comment (alexbarnsley) [inline src/renderer/components/Input/InputDelegate.vue]: I'd possibly do a `search` getter to do what is done in WalletDetails#onConfirmSelect and use that instead of piping here

File: src/renderer/components/Input/InputDelegate.vue [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+
+<script>
+import { required } from 'vuelidate/lib/validators'
+import { ButtonModal } from '@/components/Button'
+import ModalQrCodeScanner from '@/components/Modal/ModalQrCodeScanner'
+import { MenuDropdown } from '@/components/Menu'
+import Cycled from 'cycled'
+import InputField from './InputField'
+import truncate from '@/filters/truncate'
+import _ from 'lodash'
```
Comment (alexbarnsley) [inline src/renderer/components/Input/InputDelegate.vue]: It might be worth loading specifically which lodash modules we want here instead of all of them

**Comment by dated** [comment]
A `Switch Vote` functionality is something I had in mind for a future iteration 👍

**Comment by boldninja** [comment]
:waitasec: I thought that will be in this one :trollface:

**Comment by dated** [comment]
:danstop:

File: src/renderer/components/Input/InputDelegate.vue (hunk 8 above)
Comment (dated) [reply src/renderer/components/Input/InputDelegate.vue]: Something like 77f2cc78de65e3df85aaa2c43070aadef9c3ed16?

**Comment by boldninja** [comment]
Guess need to revert tier 1 :trollface:

**Comment by dated** [comment]
Just subtract from it!

**Comment by boldninja** [comment]
I'm kidding :colbert:, but if you can make switch vote in next couple of days in separate PR would be cool before next release.

File: src/renderer/components/Input/InputDelegate.vue (hunk 8 above)
Comment (alexbarnsley) [reply src/renderer/components/Input/InputDelegate.vue]: bootiful!

**Comment by alexbarnsley** [comment]
all looking good! just a message I think at the bottom of the delegate modal when already voting elsewhere, which will be replaced by the "switch votes" option when you get round to it 👌

**Comment by alexbarnsley** [comment]
I hate to say it @dated 😩 but I'm getting issues when i hit the "unvote" button in the footer. I'd screenshot but my clipboard is messed up. It's like it can't find the delegate. The modal pops up to unvote but all it says is "Unvote Delegate arkx" 🤔 will try a restart in the morning and do a screenshot then if you still need one. Double check as it may be my local data, but I believe it was working earlier

**Comment by dated** [comment]
I broke that in one of the recent commits, it's fixed now @alexbarnsley.

**Comment by dated** [comment]
There is a little issue still that's most likely caused by the api cache and the way `fetchWalletVote` works: after voting the outdated information is loaded. A proper fix for this will be possible once local storage of the transactions is implemented.

**Comment by dated** [comment]
Alternatively one could leverage the fact, that the wallet vote is stored on the wallet itself, which would eliminate the need to touch the transactions at all.

**Comment by alexbarnsley** [comment]
🤔 	well we have access to the delegate the wallet is voting for don't we. as you said in your last comment (which may be saying what i'm saying), the wallet API stores the delegate that wallet is voting for `vote: "publickey"`

**Comment by dated** [comment]
yea, we are talking about the same

**Comment by alexbarnsley** [comment]
Okay, great. Are you happy with this PR? If so, I'll give it a test

**Comment by dated** [comment]
If you are happy with me changing the vote fetching in a future iteration, then yes 😄 the API call was previously fired everytime `isAwaitingConfirmation` changed, so when a vote transaction was received in under 8 seconds it fetched the cached response, which did not include the last vote. If i'm not mistaken, right now it should fail (= show outdated information) only when all of the following happens in a timeframe of 8 seconds: open wallet -> select delegate -> input passphrase & fee -> send/receive transaction
